src/datatype.c

```diff
@@ -195,7 +247,32 @@ static jl_datatype_layout_t *jl_get_layout(uint32_t nfields,
             ptrs32[i] = pointers[i];
         }
     }
-    return flddesc;
+
+    if (__unlikely(!layoutcache_initialized)) {
+        htable_new(&layoutcache, 4096);
+        layoutcache_initialized = 1;
+    }
+
+    // Check the cache to see if this object already exists.
+    // Add to cache if not present, free temp buffer, return.
+    jl_datatype_layout_t* ret =
+            (jl_datatype_layout_t*)layoutcache_get(&layoutcache, flddesc);
+    if (ret == HT_NOTFOUND) {
```

Comment:
We have some compiler warning flags that mean you always have to cast, even for (void*), because that is the C++ rules
```suggestion
    if ((void*)ret == HT_NOTFOUND) {
```